File: packages/client/config_files.go
package client

// ConfigFile is a file projected into a service at a path.
//
// Content is absent by design: the API stores it encrypted and never returns
// it, so an agent can see that a file exists, where it mounts and who uses it,
// without the credential inside travelling any further than it must.
type ConfigFile struct {
	ID       string   `json:"id"`
	Name     string   `json:"name"`
	Path     string   `json:"path"`
	StackID  *string  `json:"stack_id"`
	Size     int      `json:"size"`
	Services []string `json:"services"`
}

func cfBase(orgID, projectID string) string {
	return "/api/v1/orgs/" + orgID + "/projects/" + projectID + "/config-files"
}

func (c *Client) ListConfigFiles(orgID, projectID string) ([]ConfigFile, error) {
	resp, err := c.do("GET", cfBase(orgID, projectID), nil)
	if err != nil {
		return nil, err
	}
	out, err := decode[struct {
		Files []ConfigFile `json:"files"`
	}](resp)
	if err != nil {
		return nil, err
	}
	return out.Files, nil
}

func (c *Client) CreateConfigFile(orgID, projectID, name, path, content string) (*ConfigFile, error) {
	type body struct {
		Name    string `json:"name"`
		Path    string `json:"path"`
		Content string `json:"content"`
	}
	resp, err := c.do("POST", cfBase(orgID, projectID), body{Name: name, Path: path, Content: content})
	if err != nil {
		return nil, err
	}
	f, err := decode[ConfigFile](resp)
	if err != nil {
		return nil, err
	}
	return &f, nil
}

func (c *Client) UpdateConfigFile(orgID, projectID, fileID, name, path, content string) (*ConfigFile, error) {
	type body struct {
		Name    string `json:"name"`
		Path    string `json:"path"`
		Content string `json:"content"`
	}
	resp, err := c.do("PATCH", cfBase(orgID, projectID)+"/"+fileID, body{Name: name, Path: path, Content: content})
	if err != nil {
		return nil, err
	}
	f, err := decode[ConfigFile](resp)
	if err != nil {
		return nil, err
	}
	return &f, nil
}

func (c *Client) DeleteConfigFile(orgID, projectID, fileID string) error {
	_, err := c.do("DELETE", cfBase(orgID, projectID)+"/"+fileID, nil)
	return err
}

func (c *Client) AttachConfigFile(orgID, projectID, fileID, serviceID string) error {
	_, err := c.do("POST", cfBase(orgID, projectID)+"/"+fileID+"/attach/"+serviceID, nil)
	return err
}

func (c *Client) DetachConfigFile(orgID, projectID, fileID, serviceID string) error {
	_, err := c.do("DELETE", cfBase(orgID, projectID)+"/"+fileID+"/attach/"+serviceID, nil)
	return err
}

File: apps/web/src/lib/api/config-files.ts
import { apiFetch } from "./core"

/**
 * A config file's `content` is never returned by the API — it is stored
 * encrypted and write-only, the same as a variable group's values. `size` is
 * carried instead so the UI can show that something is there without carrying
 * an htpasswd hash or a TLS key into the browser.
 */
export interface ApiConfigFile {
  id: string
  name: string
  path: string
  stack_id: string | null
  size: number
  /** Names of the services mounting this file. */
  services: string[]
  created_at: string
}

/** What the detail endpoint adds: service ids to link to, and a modified time. */
export interface ApiConfigFileDetail extends ApiConfigFile {
  attached_services: { id: string; name: string }[]
  updated_at: string
}

export interface ConfigFileBody {
  name: string
  path: string
  content: string
}

export const configFiles = {
  list: (orgId: string, projectId: string, token: string) =>
    apiFetch<{ files: ApiConfigFile[] }>(
      `/api/v1/orgs/${orgId}/projects/${projectId}/config-files`,
      {},
      token
    ),

  get: (orgId: string, projectId: string, fileId: string, token: string) =>
    apiFetch<ApiConfigFileDetail>(
      `/api/v1/orgs/${orgId}/projects/${projectId}/config-files/${fileId}`,
      {},
      token
    ),

  create: (orgId: string, projectId: string, body: ConfigFileBody, token: string) =>
    apiFetch<{ id: string; name: string; path: string }>(
      `/api/v1/orgs/${orgId}/projects/${projectId}/config-files`,
      { method: "POST", body: JSON.stringify(body) },
      token
    ),

  update: (orgId: string, projectId: string, fileId: string, body: ConfigFileBody, token: string) =>
    apiFetch<{ id: string; name: string; path: string }>(
      `/api/v1/orgs/${orgId}/projects/${projectId}/config-files/${fileId}`,
      { method: "PATCH", body: JSON.stringify(body) },
      token
    ),

  delete: (orgId: string, projectId: string, fileId: string, token: string) =>
    apiFetch<{ message: string }>(
      `/api/v1/orgs/${orgId}/projects/${projectId}/config-files/${fileId}`,
      { method: "DELETE" },
      token
    ),

  attach: (orgId: string, projectId: string, fileId: string, serviceId: string, token: string) =>
    apiFetch<{ message: string }>(
      `/api/v1/orgs/${orgId}/projects/${projectId}/config-files/${fileId}/attach/${serviceId}`,
      { method: "POST" },
      token
    ),

  detach: (orgId: string, projectId: string, fileId: string, serviceId: string, token: string) =>
    apiFetch<{ message: string }>(
      `/api/v1/orgs/${orgId}/projects/${projectId}/config-files/${fileId}/attach/${serviceId}`,
      { method: "DELETE" },
      token
    ),
}
